custom_components/virtual/entity.py

```diff
@@ -183,57 +185,89 @@ def _update_attributes(self):
+            self._timer_handle()
+            self._timer_handle = None
+
     def _stop(self) -> None:
-        _LOGGER.info(f"stopping {self.name}")
-        self._open_close_operation_started = None
+        _LOGGER.info(f"stopping {self.name} at position {self._current_position}")
+
+        self._cancel_timer()
+
         self._target_position = None
+        self._positions_per_tick = None
         self._attr_is_opening = False
         self._attr_is_closing = False
 
-        if self._current_position > 0:
-            self._attr_is_closed = False
+        self._attr_is_closed = (self._current_position == 0)
+
+        self.async_write_ha_state()
+
+    def _set_direction_flags(self, target_position: int) -> None:
+        """Set opening/closing flags based on target position."""
+        if target_position < self._current_position:
+            self._attr_is_closing = True
+            self._attr_is_opening = False
         else:
-            self._attr_is_closed = True
+            self._attr_is_opening = True
+            self._attr_is_closing = False
 
         self.async_write_ha_state()
 
-    def _set_position(self, position) -> None:
+    def _set_position(self, position: int) -> None:
         _LOGGER.info(f"setting {self.name} position {position}")
 
+        self._cancel_timer()
+
+        position = max(0, min(100, int(position)))
+
         self._target_position = position
+
         if self._target_position == self._current_position:
             return
-        elif self._target_position < self._current_position:
+
+        if self._open_close_tick > self._open_close_duration:
+            _LOGGER.warning(f"Tick duration {self._open_close_tick} > total duration {self._open_close_duration}, capping to {self._open_close_duration}")
+            self._open_close_tick = self._open_close_duration
+
+        if self._open_close_duration == 0:
+            # Transition through opening/closing state for automations
+            self._set_direction_flags(self._target_position)
+
+            # Immediately set final state
+            self._current_position = self._target_position
             self._attr_is_opening = False
-            self._attr_is_closing = True
-        else:
-            self._attr_is_opening = True
             self._attr_is_closing = False
+            self._attr_is_closed = (self._current_position == 0)
+            self._target_position = None
 
-        self.async_write_ha_state()
-        self._open_close_operation_started = datetime.now()
-        async_call_later(self.hass, self._open_close_tick, self._update_position)
+            self.async_schedule_update_ha_state(force_refresh=True)
+            return
+
+        distance = abs(self._target_position - self._current_position)
+        movement_duration = (distance / 100.0) * self._open_close_duration
+        total_ticks = max(1, int(movement_duration / self._open_close_tick))
+        self._positions_per_tick = min(1, round(distance / total_ticks))
```

Comment:
I think the `min()` call is wrong. I thought it needed to be `max()` but I think nothing is better. 

If the open takes 200 seconds and tick is set to 1  then position per tick is 0.5
If the open takes 10 seconds and tick is set to 1 then position per tick is 10.